packages/react-components/react-positioning/src/middleware/maxSize.ts

```diff
@@ -1,40 +1,69 @@
 import { size } from '@floating-ui/dom';
 import type { Middleware } from '@floating-ui/dom';
-import type { PositioningOptions } from '../types';
+import type { NormalizedAutoSize, PositioningOptions } from '../types';
 import { getBoundary } from '../utils/getBoundary';
 export interface MaxSizeMiddlewareOptions extends Pick<PositioningOptions, 'overflowBoundary'> {
   container: HTMLElement | null;
 }
 
-export function maxSize(autoSize: PositioningOptions['autoSize'], options: MaxSizeMiddlewareOptions): Middleware {
+/**
+ * floating-ui `size` middleware uses floating element's height/width to calculate available height/width.
+ * This middleware only runs once per lifecycle, resetting styles applied by maxSize from previous lifecycle.
+ * Then floating element's original size is restored and `size` middleware can calculate available height/width correctly.
+ */
+export const resetMaxSize = (normalizedAutoSize: NormalizedAutoSize): Middleware => ({
+  name: 'resetMaxSize',
+  fn({ middlewareData: { maxSizeAlreadyReset }, elements }) {
+    if (maxSizeAlreadyReset) {
+      return {};
+    }
+
+    const { applyMaxWidth, applyMaxHeight } = normalizedAutoSize;
+    if (applyMaxWidth) {
+      elements.floating.style.removeProperty('box-sizing');
+      elements.floating.style.removeProperty('max-width');
+      elements.floating.style.removeProperty('width');
+    }
+    if (applyMaxHeight) {
+      elements.floating.style.removeProperty('box-sizing');
+      elements.floating.style.removeProperty('max-height');
+      elements.floating.style.removeProperty('height');
+    }
+
+    return {
+      data: { maxSizeAlreadyReset: true },
+      reset: { rects: true },
+    };
+  },
+});
+
+export function maxSize(normalizedAutoSize: NormalizedAutoSize, options: MaxSizeMiddlewareOptions): Middleware {
   const { container, overflowBoundary } = options;
   return size({
     ...(overflowBoundary && { altBoundary: true, boundary: getBoundary(container, overflowBoundary) }),
     apply({ availableHeight, availableWidth, elements, rects }) {
-      if (autoSize) {
+      const { applyMaxWidth, applyMaxHeight } = normalizedAutoSize;
+
+      if (applyMaxWidth) {
```

Comment:
nit you can pare down this code by parametrizing `width`/`height` - `dimension`, `x`/`y` - `axis` and `availableWidth`/`availableHeight` - `size`

based on `applyMaxWidth` or `applyMaxHeigh`

Reply:
You actually apply the CSS properties in the same way for either case